internal/tcptunnel/tcptunnel.go

```diff
@@ -0,0 +1,211 @@
+// Package tcptunnel contains an implementation of a TCP tunnel via HTTP Connect.
+package tcptunnel
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/pomerium/pomerium/internal/authclient"
+	"github.com/pomerium/pomerium/internal/cliutil"
+	"github.com/pomerium/pomerium/internal/log"
+
+	backoff "github.com/cenkalti/backoff/v4"
+)
+
+// A Tunnel represents a TCP tunnel over HTTP Connect.
+type Tunnel struct {
+	cfg  *config
+	auth *authclient.AuthClient
+}
+
+// New creates a new Tunnel.
+func New(options ...Option) *Tunnel {
+	cfg := getConfig(options...)
+	return &Tunnel{
+		cfg:  cfg,
+		auth: authclient.New(authclient.WithTLSConfig(cfg.tlsConfig)),
+	}
+}
+
+// RunListener runs a network listener on the given address. For each
+// incoming connection a new TCP tunnel is established via Run.
+func (tun *Tunnel) RunListener(ctx context.Context, listenerAddress string) error {
+	li, err := net.Listen("tcp", listenerAddress)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = li.Close() }()
+	log.Info().Msg("tcptunnel: listening on " + li.Addr().String())
+
+	go func() {
+		<-ctx.Done()
+		_ = li.Close()
+	}()
+
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxElapsedTime = 0
+
+	for {
+		conn, err := li.Accept()
+		if err != nil {
+			// canceled, so ignore the error and return
+			if ctx.Err() != nil {
+				return nil
+			}
+
+			if nerr, ok := err.(net.Error); ok && nerr.Temporary() {
+				log.Warn().Err(err).Msg("tcptunnel: temporarily failed to accept local connection")
+				select {
+				case <-time.After(bo.NextBackOff()):
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				continue
+			}
+			return err
+		}
+		bo.Reset()
+
+		go func() {
+			defer func() { _ = conn.Close() }()
+
+			err := tun.Run(ctx, conn)
+			if err != nil {
+				log.Error().Err(err).Msg("tcptunnel: error serving local connection")
+			}
+		}()
+	}
+}
+
+// Run establishes a TCP tunnel via HTTP Connect and forwards all traffic from/to local.
+func (tun *Tunnel) Run(ctx context.Context, local io.ReadWriter) error {
+	rawJWT, err := tun.cfg.jwtCache.LoadJWT(tun.jwtCacheKey())
+	switch {
+	case err == nil:
+	case errors.Is(err, cliutil.ErrExpired):
```

Comment:
Can you explain this switch statement?

Reply:
I read this as "If there is no error, or is one I expect, continue with tun.Run"